src/enclave/Enclave/IntegrityUtils.cpp

```diff
@@ -0,0 +1,172 @@
+#include "IntegrityUtils.h"
+
+void init_log(const tuix::EncryptedBlocks *encrypted_blocks) {
+  // Add past entries to log first
+  std::vector<LogEntry> past_log_entries;
+  auto curr_entries_vec = encrypted_blocks->log()->curr_entries();
+  auto past_entries_vec = encrypted_blocks->log()->past_entries();
+
+  for (uint32_t i = 0; i < past_entries_vec->size(); i++) {
+    auto entry = past_entries_vec->Get(i);
+    std::string ecall = entry->ecall()->str();
+    int snd_pid = entry->snd_pid();
+    int rcv_pid = entry->rcv_pid();
+    if (rcv_pid == -1) { // Received by PID hasn't been set yet
+      rcv_pid = EnclaveContext::getInstance().get_pid();
+    }
+    int job_id = entry->job_id();
+    EnclaveContext::getInstance().append_past_log_entry(ecall, snd_pid, rcv_pid, job_id);
+
+    // Initialize log entry object
+    LogEntry le;
+    le.ecall = ecall;
+    le.snd_pid = snd_pid;
+    le.rcv_pid = rcv_pid;
+    le.job_id = job_id;
+    past_log_entries.push_back(le);
+  }
+
+  if (curr_entries_vec->size() > 0) {
+    verify_log(encrypted_blocks, past_log_entries);
+  }
+
+  // Master list of mac lists of all input partitions
+  std::vector<std::vector<std::vector<uint8_t>>> partition_mac_lsts;
+
+  // Check that each input partition's global_mac is indeed a HMAC over the mac_lst
+  for (uint32_t i = 0; i < curr_entries_vec->size(); i++) {
+    auto input_log_entry = curr_entries_vec->Get(i);
+
+    // Copy over the global mac for this input log entry
+    uint8_t global_mac[SGX_AESGCM_MAC_SIZE];
+    memcpy(global_mac, input_log_entry->global_mac()->data(), SGX_AESGCM_MAC_SIZE);
```

Comment:
Why is this memcpy necessary? Can you access the deserialized flatbuffers data directly?

Reply:
Fixed